projects/dendrites/permutedMNIST/experiments/cluster.py

```diff
@@ -0,0 +1,160 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2021, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+Experiment file that runs dendritic networks which infer the context vector via 1)
+applying a clustering procedure during training, and 2) inferring the closest prototype
+as context during inference.
+"""
+
+import os
+from copy import deepcopy
+
+import numpy as np
+import ray.tune as tune
+import torch
+import torch.nn.functional as F
+
+from nupic.research.frameworks.dendrites import DendriticMLP
+from nupic.research.frameworks.dendrites.dendrite_cl_experiment import (
+    DendriteContinualLearningExperiment,
+)
+from nupic.research.frameworks.pytorch.datasets import PermutedMNIST
+from nupic.research.frameworks.vernon import mixins
+
+
+class CentroidExperiment(mixins.RezeroWeights,
+                         mixins.CentroidContext,
+                         mixins.PermutedMNISTTaskIndices,
+                         DendriteContinualLearningExperiment):
+    pass
+
+
+CENTROID_CLUSTER_5 = dict(
+    experiment_class=CentroidExperiment,
+    num_samples=8,
+
+    # Results path
+    local_dir=os.path.expanduser("~/nta/results/experiments/dendrites"),
+
+    dataset_class=PermutedMNIST,
+    dataset_args=dict(
+        num_tasks=5,
+        root=os.path.expanduser("~/nta/results/data/"),
+        download=False,  # Change to True if running for the first time
+        seed=42,
+    ),
+
+    model_class=DendriticMLP,  # CentroidDendriticMLP does not affect accuracy..??
+    model_args=dict(
+        input_size=784,
+        output_size=10,  # Single output head shared by all tasks
+        hidden_sizes=[2048, 2048],
+        num_segments=5,
+        dim_context=256,
+        kw=True,
+        kw_percent_on=0.1,
+        dendrite_weight_sparsity=0.0,
+        weight_sparsity=0.5,
+        context_percent_on=0.05,
+    ),
+
+    centroid_context_args=dict(infer_while_training=True),
+
+    batch_size=256,
+    val_batch_size=512,
+    epochs=5,
```

Comment:
we used to do 3 epochs before no ?

Reply:
ah ok I see below it's updated as a function of cluster size. Is it for optimization ?